### PR TITLE
Add Default UI-Schema Generator

### PR DESCRIPTION
**Pull request by sdirix**

### CODE REVIEW

**Comment by koegel** [comment]
Please review and merge.

File: app/js/DefaultUISchemaController.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                "type": "string"
+            },
+            "age": {
+                "type": "integer"
+            }
+        }
+    };
+    $scope.uiSchema = undefined;
+
+    $scope.usersSchema = {
```
Comment (edgarmueller) [inline app/js/DefaultUISchemaController.js]: Is this actually used anywhere?

File: app/js/DefaultUISchemaController.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        "type": "array",
+        "items": $scope.schema
+    };
+    $scope.usersUiSchema = {
```
Comment (edgarmueller) [inline app/js/DefaultUISchemaController.js]: Applies to usersUiSchema as well.

File: app/templates/defaultui.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        </div>
+    </div>
+
+    <div class="panel panel-default">
+        <div class="panel-heading">
+            <h3>JSONForms generated default Forms</h3>
+        </div>
+        <div class="panel-body">
+            <jsonforms schema="schema" ui-schema="uiSchema" data="data"></jsonforms>
+            <!--<jsonforms schema="usersSchema" ui-schema="usersUiSchema" data="users"></jsonforms>-->
```
Comment (edgarmueller) [inline app/templates/defaultui.html]: Line might be removed.

File: js/directives.ts [hunk 4]
```diff
@@ -93,7 +120,13 @@ class JsonFormsDiretiveController {
             return this.$scope.asyncUiSchema();
         }
 
-        throw new Error("Either the 'ui-schema' or the 'async-ui-schema' attribute must be specified.");
+        // throw new Error("Either the 'ui-schema' or the 'async-ui-schema' attribute must be specified.");
```
Comment (edgarmueller) [inline js/directives.ts]: Remove line.

File: js/services.ts [hunk 5]
```diff
@@ -135,11 +139,130 @@ module jsonforms.services {
...
+            var uiSchemaElements = [];
+            this.generateUISchema(jsonSchema, uiSchemaElements, "#", "");
+
+            console.log("generated schema: " + JSON.stringify(uiSchemaElements[0]))
+
+            return uiSchemaElements[0];
+        };
+
+        private generateUISchema = (jsonSchema:any, schemaElements:IUISchemaElement[], currentRef:string, schemaName:string):any =>{
+            if(!jsonSchema.type){
```
Comment (edgarmueller) [inline js/services.ts]: As a side note: JSON schema does not necessarily require the type keyword to be defined on a schema, for instance, the presence of the properties keyword suffices in order to derive the object type.

File: dist/js/jsonforms.js [hunk 6]
```diff
@@ -411,174 +480,271 @@ var app = angular.module('jsonForms.verticalLayout', ['jsonForms.services']);
...
+                                verticalLayout.elements.push(label);
+                            }
+                            // traverse properties
+                            if (!jsonSchema.properties) {
+                                // If there are no properties return
+                                return;
+                            }
+                            var nextRef = currentRef + '/' + "properties";
+                            for (var property in jsonSchema.properties) {
+                                if (property === "id") {
```
Comment (edgarmueller) [inline dist/js/jsonforms.js]: There are more meta-keywords that all should be ignored here (see spec)

File: dist/js/jsonforms.js [hunk 7]
```diff
@@ -411,174 +480,271 @@ var app = angular.module('jsonForms.verticalLayout', ['jsonForms.services']);
...
+                    };
+                };
+                //1. split on uppercase letters
+                //2. transform uppercase letters to lowercase
+                //3. transform first letter uppercase
+                this.beautify = function (text) {
+                    if (text && text.length > 0) {
+                        var textArray = text.split(/(?=[A-Z])/).map(function (x) { return x.toLowerCase(); });
+                        textArray[0] = textArray[0].charAt(0).toUpperCase() + textArray[0].slice(1);
+                        text = textArray.join(' ');
```
Comment (edgarmueller) [inline dist/js/jsonforms.js]: return directly instead of re-assigning param

File: js/services.ts [hunk 8]
```diff
@@ -135,11 +139,130 @@ module jsonforms.services {
...
+        };
+
+        //1. split on uppercase letters
+        //2. transform uppercase letters to lowercase
+        //3. transform first letter uppercase
+        private beautify = (text: string): string => {
+            if(text && text.length > 0){
+                var textArray = text.split(/(?=[A-Z])/).map((x)=>{return x.toLowerCase()});
+                textArray[0] = textArray[0].charAt(0).toUpperCase() + textArray[0].slice(1);
+                text = textArray.join(' ');
```
Comment (edgarmueller) [inline js/services.ts]: return directly instead of re-assigning param

File: js/services.ts [hunk 9]
```diff
@@ -135,11 +139,130 @@ module jsonforms.services {
...
+
+                    // traverse properties
+                    if (!jsonSchema.properties) {
+                        // If there are no properties return
+                        return;
+                    }
+
+                    var nextRef:string = currentRef + '/' + "properties";
+                    for (var property in jsonSchema.properties) {
+                        if(property === "id"){
```
Comment (edgarmueller) [inline js/services.ts]: There are more meta-keywords that all should be ignored here (see spec)

File: typings/schemas/jsonschema.d.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
+interface SchemaElement{
+    type: string;
+}
+
+interface SchemaObject extends SchemaElement{
+    //TODO: specify properties
+    properties: any;
+    additionalProperties?: boolean;
```
Comment (edgarmueller) [inline typings/schemas/jsonschema.d.ts]: additionalProperties may also be of type object

**Comment by edgarmueller** [comment]
I've added some notes which you might wanna have a look at. Otherwise this looks good. I will merge this as soon as you have looked at the comments.

File: app/js/DefaultUISchemaController.js (hunk 1 above)
Comment (sdirix) [inline app/js/DefaultUISchemaController.js]: Copied over from LocalController. Should probably be removed there too.

**Comment by sdirix** [comment]
Thanks for your comments ;)

**Pull request merged**